python/rmm/rmm/_lib/memory_resource.py

```diff
@@ -0,0 +1,44 @@
+# Copyright (c) 2018-2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from rmm.pylibrmm.memory_resource import (  # noqa: F401
+    BinningMemoryResource,
```

Comment:
These classes are `cdef`'d, why are they importable from python? Should they be `cpdef`'d?

Reply:
A `cdef class` is a normal python class that can also hold C properties. (https://cython.readthedocs.io/en/latest/src/tutorial/cdef_classes.html)

`cpdef` is for (predominantly) functions which should have a "fast-path" calling route from typed code in cython, and also a "slow-path" python-like calling route.

Reply:
So its only `cpdef` functions that can be called from C and Python?

Reply:
Not exactly.
- cdef functions can only be called from Cython and in typed contexts (i.e. you cannot pass arbitrary Python objects to them, you need to pass objects that have already been cdefed appropriately themselves).
- def functions can be called from Python or Cython. They are pure Python functions, so all arguments are treated as normal Python objects.
- cpdef functions can be called from Python or Cython. When called from Python, they behave like def functions. When called from Cython with untyped arguments, they also behave like def functions. When called from Cython with typed arguments, they behave like cdef functions.

There's more information [here](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#python-functions-vs-c-functions).

Reply:
Thanks for the clarification (def and cpdef make sense to me). Can you give an example of a "typed context"?

Reply:
Can a typed context be in a *.py file?

Reply:
Is it the body of cpdef and cdef functions/methods?

Reply:
```
cdef class A:
    cpdef class f(self):
        pass

A().f()  # Called like a Python def function
cdef A a = A()
a.f() # Called like a cdef function.
```

So no, a typed context cannot be in a *.py file, or at least not in the way that we use Cython. With newer versions of Cython you can use py files and compile them too, but the point is that you have to use Cython-specific syntax.